physical/gcs/gcs.go

```diff
@@ -88,6 +75,53 @@ func NewGCSBackend(conf map[string]string, logger log.Logger) (physical.Backend,
 	return &g, nil
 }
 
+func NewGCSClient(conf map[string]string, logger log.Logger) (*storage.Client, error) {
+	var client *storage.Client
+	ctx := context.Background()
+
+	// Use "Application Default Credentials" to auth to GCS
+	useADC, ok := conf["use_application_default_credentials"]
+	if ok && useADC != "" {
+		useApplicationDefaultCredentials, err := strconv.ParseBool(useADC)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed parsing use_application_default_credentials parameter: {{err}}", err)
+		}
+		if logger.IsDebug() {
+			logger.Debug("physical/gcs: config use_application_default_credentials set", "use_application_default_credentials", useApplicationDefaultCredentials)
+		}
+		if useApplicationDefaultCredentials {
+			client, err := storage.NewClient(ctx)
+			if err != nil {
+				return nil, errwrap.Wrapf("error establishing storage client with application default credentials: {{err}}", err)
+			}
+			return client, nil
+		}
+	}
+
+	// path to service account JSON file
+	credentialsFile := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
```

Comment:
If we have the option to use application default credentials, we might as while get rid of the code to read `GOOGLE_APPLICATION_CREDENTIALS` env variable here, since that is the first place that ADC looks for credentials. Any issue w/ this @jefferai? I think as long as we point to ADC docs in the vault docs this should be fine.

Reply:
The only caveat is that we'd have to put this is a breaking change in the Vault release notes, since GCS storage users would have to turn on the `use_application_default_credentials` flag if they are using `GOOGLE_APPLICATION_CREDENTIALS`  env variable.

Reply:
You might want to keep `GOOGLE_APPLICATION_CREDENTIALS` because this is the standard way to configure google SDK credentials when you are *not* running in GCP. It could be considered redundant to have both this and `credentials_file` as a vault config line, but a lot of applications follow the same pattern of using both env vars and configs. Based on that, I see no reason to go out of our way to remove it, and thus introduce a breaking change.

Reply:
Yeah, I think you're right, we should just keep it. The benefit of removing it is minimal and it's a breaking change.